tests/test_hll.erl

```diff
@@ -0,0 +1,179 @@
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+%%% @doc r_t to verify changes to serialization/metadata/changes
+%%       across dt upgrades
+
+-module(test_hll).
+
+-export([confirm/0]).
+
+-include_lib("eunit/include/eunit.hrl").
+
+-define(HLL_TYPE1, <<"hlls1">>).
+-define(HLL_TYPE2, <<"hlls2">>).
+-define(BUCKET1, {?HLL_TYPE1, <<"testbucket1">>}).
+-define(BUCKET2, {?HLL_TYPE2, <<"testbucket2">>}).
+-define(DEFAULT_P, 14).
+-define(SET_P, 16).
+-define(P_SETTING, hll_precision).
+-define(KEY, <<"flipit&reverseit">>).
+-define(CONFIG,
+        [
+         {riak_core,
+          [{ring_creation_size, 8},
+           {anti_entropy_build_limit, {100, 1000}},
+           {anti_entropy_concurrency, 8}]
+         }
+        ]).
+
+confirm() ->
+    %% Configure cluster.
+    Nodes = rt:build_cluster(4, ?CONFIG),
+
+    NodeRand1 = rt:select_random(Nodes),
+    NodeRand2 = rt:select_random(Nodes),
+
+    %% Create PB connection.
+    PBC = rt:pbc(NodeRand1),
+    riakc_pb_socket:set_options(PBC, [queue_if_disconnected]),
+
+    %% Create HTTPC connection.
+    HttpC = rt:httpc(NodeRand2),
+
+    rt:create_and_activate_bucket_type(NodeRand1,
+                                       ?HLL_TYPE1,
+                                       [{datatype, hll},
+                                        {hll_precision, ?SET_P}]),
+
+    rt:create_and_activate_bucket_type(NodeRand1,
+                                       ?HLL_TYPE2,
+                                       [{datatype, hll}]),
+
+    pb_tests(PBC, riakc_pb_socket, ?BUCKET1, Nodes),
+    http_tests(HttpC, rhc, ?BUCKET2, Nodes),
+
+    %% Stop PB connection.
+    riakc_pb_socket:stop(PBC),
+
+    %% Clean cluster
+    rt:clean_cluster(Nodes),
+
+    pass.
+
+http_tests(C, CMod, Bucket, Nodes) ->
+    lager:info("HTTP CLI TESTS: Create new Hll DT"),
+    S0 = riakc_hll:new(),
+
+    add_tests(C, CMod, Bucket, S0),
+
+    HllSet0 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_test(C, CMod, Bucket, ?DEFAULT_P, HllSet0),
+
+    partition_write_heal(C, CMod, Bucket, Nodes),
+
+    HllSet1 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_invalid_test(C, CMod, Bucket, ?DEFAULT_P - 1,
+                                       HllSet1),
+
+    ok.
+
+pb_tests(C, CMod, Bucket, Nodes) ->
+    lager:info("PB CLI TESTS: Create new Hll DT"),
+    S0 = riakc_hll:new(),
+
+    add_tests(C, CMod, Bucket, S0),
+
+    HllSet0 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_test(C, CMod, Bucket, ?SET_P, HllSet0),
+
+    partition_write_heal(C, CMod, Bucket, Nodes),
+
+    HllSet1 = get_hll(C, CMod, Bucket),
+    check_precision_and_reduce_invalid_test(C, CMod, Bucket, ?SET_P - 1, HllSet1),
+
+    ok.
+
+add_tests(C, CMod, Bucket, NewS) ->
```

Comment:
OH I SEE, the code explicitly expects this to be a New Set since the first thing it does is add an element and check the cardinality is 1. I get it now.